pkg/containerruntimediscovery/containerddiscoverer.go

```diff
@@ -0,0 +1,191 @@
+	if err != nil {
+		return models.ContainerImageInfo{}, fmt.Errorf("failed to load image spec: %w", err)
+	}
+
+	// NOTE(sambetts) We can not use image.Size as it gives us the size of
+	// the compressed layers and not the real size of the content.
+	snapshotter := cd.client.SnapshotService(containerd.DefaultSnapshotter)
+	size, err := imgutil.UnpackedImageSize(ctx, snapshotter, image)
+	if err != nil {
+		return models.ContainerImageInfo{}, fmt.Errorf("unable to determine size for image: %w", err)
+	}
+
+	repoTags, repoDigests := ParseImageReferences([]string{image.Name()})
+
+	return models.ContainerImageInfo{
+		ImageID:      id,
+		Architecture: utils.PointerTo(imageSpec.Architecture),
+		Labels:       convertTags(imageSpec.Config.Labels),
+		RepoTags:     &repoTags,
+		RepoDigests:  &repoDigests,
+		ObjectType:   "ContainerImageInfo",
+		Os:           utils.PointerTo(imageSpec.OS),
+		Size:         utils.PointerTo(int(size)),
+	}, nil
+}
+
+// ParseImageReferences parses a list of arbitrary image references and returns
+// the repotags and repodigests.
+func ParseImageReferences(refs []string) ([]string, []string) {
+	var tags, digests []string
+	for _, ref := range refs {
+		parsed, err := reference.ParseAnyReference(ref)
+		if err != nil {
+			continue
+		}
+		if _, ok := parsed.(reference.Canonical); ok {
+			digests = append(digests, parsed.String())
+		} else if _, ok := parsed.(reference.Tagged); ok {
+			tags = append(tags, parsed.String())
+		}
+	}
+	return tags, digests
+}
+
+func (cd *ContainerdDiscoverer) Containers(ctx context.Context) ([]models.ContainerInfo, error) {
+	containers, err := cd.client.Containers(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("unable to list containers: %w", err)
+	}
+
+	result := make([]models.ContainerInfo, len(containers))
+	for i, container := range containers {
+		// Get container info
+		info, err := cd.getContainerInfo(ctx, container)
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert container to ContainerInfo: %w", err)
+		}
+		result[i] = info
+	}
+	return result, nil
+}
+
+func (cd *ContainerdDiscoverer) getContainerInfo(ctx context.Context, container containerd.Container) (models.ContainerInfo, error) {
+	id := container.ID()
+
+	labels, err := container.Labels(ctx)
+	if err != nil {
+		return models.ContainerInfo{}, fmt.Errorf("unable to get labels for container %s: %w", id, err)
+	}
+	// If this doesn't exist then use empty string as the name. Containerd
+	// doesn't have the concept of a Name natively.
+	name := labels[k8slabels.ContainerName]
+
+	info, err := container.Info(ctx)
+	if err != nil {
+		return models.ContainerInfo{}, fmt.Errorf("unable to get info for container %s: %w", id, err)
+	}
+	createdAt := info.CreatedAt
+
+	image, err := container.Image(ctx)
+	if err != nil {
+		return models.ContainerInfo{}, fmt.Errorf("unable to get image from container %s: %w", id, err)
+	}
+
+	imageInfo, err := cd.getContainerImageInfo(ctx, image)
```

Comment:
It make more sense to me here to look for the image by id in the previously found images, rather than recalculate the data. WDYT?

Reply:
yeah, I mentioned this to @chrisgacsal I think we need to change the API model so that we just store the ID and then we can look it up later. However it would mean that if we find a container and we are missing the ContainerImageInfo asset that we'll be missing a lot of information.

Reply:
I meant look in the imageIDs that you found as part of this discovery session. 
```
images := ListImages()
containers := ListContainers(images)
```

and then in ListContainers, we can match imageId to image data in images.

Reply:
that will require holding every image that we found during the discovery run in memory in the orchestrator as there is no access to the API server as part of the provider side of the discovery code. If we change the API model so that the container's just store a reference / imageID, then we can make the link in API / UI layer instead of reimplementing it in each provider.